src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/HandleCollector.cs

```diff
@@ -0,0 +1,129 @@
+        {
+            int initial = 10;
+            int max = 20;
+            HandleCollector collector = new HandleCollector(null, initial, max);
+
+            Assert.Equal(0, collector.Count);
+            Assert.Equal(initial, collector.InitialThreshold);
+            Assert.Equal(max, collector.MaximumThreshold);
+            Assert.Equal(string.Empty, collector.Name);
+        }
+
+        [Fact]
+        public static void EmptyRemoval()
+        {
+            HandleCollector collector = new HandleCollector("EmptyRemoval", 10);
+            Assert.Throws<InvalidOperationException>(() => collector.Remove());
+        }
+
+        [Fact]
+        [OuterLoop]
+        public static void CountOverflow()
+        {
+            HandleCollector collector = new HandleCollector("CountOverflow", int.MaxValue);
+            for (int i = 0; i < Int32.MaxValue; i++)
+            {
+                collector.Add();
+            }
+
+            Assert.Throws<InvalidOperationException>(() => collector.Add());
+        }
+
+        [Fact]
+        public static void TestHandleCollector()
+        {
+            Tuple<int, int, int> intialGcState = new Tuple<int, int, int>(
+                GC.CollectionCount(0),
+                GC.CollectionCount(1),
+                GC.CollectionCount(2));
+
+            for (int i = 0; i < LowLimitSize + 1; ++i)
+            {
+                TestLowLimitHandle t = new TestLowLimitHandle();
+            }
+
+            Tuple<int, int, int> postLowLimitState = new Tuple<int, int, int>(
+                GC.CollectionCount(0),
+                GC.CollectionCount(1),
+                GC.CollectionCount(2));
+
+            Assert.True(intialGcState.Item1 + intialGcState.Item2 + intialGcState.Item3 < postLowLimitState.Item1 + postLowLimitState.Item2 + postLowLimitState.Item3, "Low limit handle did not trigger a GC");
+
+            for (int i = 0; i < HighLimitSize + 10; ++i)
+            {
+                TestHighLimitHandle t = new TestHighLimitHandle();
+            }
+
+            Tuple<int, int, int> postHighLimitState = new Tuple<int, int, int>(
+                GC.CollectionCount(0),
+                GC.CollectionCount(1),
+                GC.CollectionCount(2));
+
+            Assert.True(postLowLimitState.Item1 + postLowLimitState.Item2 + postLowLimitState.Item3 < postHighLimitState.Item1 + postHighLimitState.Item2 + postHighLimitState.Item3, "High limit handle did not trigger a GC");
+        }
+
+        private sealed class TestLowLimitHandle
+        {
+            private static HandleCollector s_lowLimitCollector = new HandleCollector("LowLimit.Collector", LowLimitSize);
+
+            internal TestLowLimitHandle()
+            {
+                s_lowLimitCollector.Add();
+                GC.KeepAlive(this);
+            }
+
+            ~TestLowLimitHandle()
+            {
+                s_lowLimitCollector.Remove();
+            }
+        }
+
+        private sealed class TestHighLimitHandle
```

Comment:
You could collapse these two Test*LimitHandle implementations by passing the relevant HandleCollector into the ctor.